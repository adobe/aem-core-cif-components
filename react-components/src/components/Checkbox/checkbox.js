/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { Component, Fragment } from 'react';
import { bool, node, shape, string } from 'prop-types';
import { BasicCheckbox, asField } from 'informed';

import { Message } from '../Field';
import Icon from '../Icon';
import { Check as CheckIcon } from 'react-feather';
import classes from './checkbox.css';

export class Checkbox extends Component {
    static propTypes = {
        classes: shape({
            icon: string,
            input: string,
            label: string,
            message: string,
            root: string
        }),
        field: string.isRequired,
        fieldState: shape({
            value: bool
        }).isRequired,
        id: string,
        label: node.isRequired,
        message: node
    };

    render() {
        const { fieldState, id, label, message, ...rest } = this.props;
        const { value: checked } = fieldState;

        return (
            <Fragment>
                <label className={classes.root} htmlFor={id}>
                    <span className={classes.icon}>{checked && <Icon src={CheckIcon} size={18} />}</span>
                    <BasicCheckbox {...rest} className={classes.input} fieldState={fieldState} id={id} />
                    <span className={classes.label}>{label}</span>
                </label>
                <Message fieldState={fieldState}>{message}</Message>
            </Fragment>
        );
    }
}

export default asField(Checkbox);
