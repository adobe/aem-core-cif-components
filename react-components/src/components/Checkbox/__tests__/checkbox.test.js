/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { Form } from 'informed';
import { render } from '@testing-library/react';

import Checkbox from '../checkbox';

const field = 'a';
const label = 'b';
const id = 'c';

const props = { field, label };

describe('<Checkbox>', () => {
    it('renders the correct tree', () => {
        const { asFragment } = render(
            <Form>
                <Checkbox {...props} id={id} />
            </Form>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('applies `props.id` to both label and input', () => {
        const { container, getByRole } = render(
            <Form>
                <Checkbox {...props} id={id} />
            </Form>
        );

        const labelInstance = container.querySelector('label');
        const checkboxInstance = getByRole('checkbox');

        expect(checkboxInstance.id).toBe(id);
        expect(labelInstance.htmlFor).toBe(id);
        expect(labelInstance.id).toBe('');
    });

    it('applies `checked` based on `initialValue`', () => {
        const { getAllByRole } = render(
            <Form>
                <Checkbox {...props} field={'a.x'} initialValue={true} />
                <Checkbox {...props} field={'a.y'} initialValue={false} />
            </Form>
        );

        const [x, y] = getAllByRole('checkbox');

        expect(x.checked).toBe(true);
        expect(y.checked).toBe(false);
    });
});
