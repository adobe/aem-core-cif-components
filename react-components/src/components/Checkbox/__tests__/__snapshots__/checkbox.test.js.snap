// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> renders the correct tree 1`] = `
<DocumentFragment>
  <form>
    <label
      class="root"
      for="c"
    >
      <span
        class="icon"
      />
      <input
        class="input"
        id="c"
        name="a"
        type="checkbox"
      />
      <span
        class="label"
      >
        b
      </span>
    </label>
    <p
      class="root"
    />
  </form>
</DocumentFragment>
`;
